docs/config.md
# Config

`boltz-lnd` can be configured via CLI arguments, and a TOML config file. By default, the config file is called `boltz.toml` and located in the default data directory of `boltz-lnd` (`/home/<user>/.boltz-lnd` on Linux).

## Example

```toml
# Path the the log file
logfile = ""

# Prefix for all log messages.
# Useful in cases two boltz-lnd instances (one for BTC and LTC) are running in a single Docker container  
logprefix = "[BTC] "

[BOLTZ]
# By default the daemon automatically connects to the official Boltz instance for the network the LND is on
# This value is used to override that
url = "https://testnet.boltz.exchange/api"

[DATABASE]
# Path to the SQLite database file 
path = "/home/michael/test.db"

[LND]
# Host of the gRPC interface of LND
host = "127.0.0.1"

# Port of the gRPC interface of LND
port = 10009

# Path to a macaroon file of LND
# The daemon needs to have permission to read various endpoints, generate addresses and pay invoices 
macaroon = ""

# Path to the TLS certificate of LND
certificate = ""

[RPC]
# Host of the gRPC interface
host = "127.0.0.1"

# Port of the gRPC interface 
port = 9002

# Whether the REST proxy for the gRPC interface should be disabled
restDisabled = false

# Host of the REST proxy
restHost = "127.0.0.1"

# Port of the REST proxy
restPort = 9003

# Path to the TLS cert for the gRPC and REST interface
tlsCert = ""

# Path to the TLS private key for the gRPC and REST interface
tlsKey = ""

# Whether the macaroon authentication for the gRPC and REST interface should be disabled
noMacaroons = false

# Path to the admin macaroon for the gRPC and REST interface
adminMacaroonPath = ""

# Path to the read only macaroon for the gRPC and REST interface
readOnlyMacaroonPath = ""
```


docs/configuration.md
# Configuration

`boltz-lnd` can be configured via CLI arguments, and a TOML config file. By default, the config file is called `boltz.toml` and located in the default data directory of `boltz-lnd` (`/home/<user>/.boltz-lnd` on Linux).

## Example

```toml
# Path the the log file
logfile = ""

# Prefix for all log messages.
# Useful in cases two boltz-lnd instances (one for BTC and LTC) are running in a single Docker container  
logprefix = "[BTC] "

[BOLTZ]
# By default the daemon automatically connects to the official Boltz instance for the network the LND is on
# This value is used to override that
url = "https://testnet.boltz.exchange/api"

[DATABASE]
# Path to the SQLite database file 
path = "/home/michael/test.db"

[LND]
# Host of the gRPC interface of LND
host = "127.0.0.1"

# Port of the gRPC interface of LND
port = 10009

# Path to a macaroon file of LND
# The daemon needs to have permission to read various endpoints, generate addresses and pay invoices 
macaroon = ""

# Path to the TLS certificate of LND
certificate = ""

[RPC]
# Host of the gRPC interface
host = "127.0.0.1"

# Port of the gRPC interface 
port = 9002

# Whether the REST proxy for the gRPC interface should be disabled
restDisabled = false

# Host of the REST proxy
restHost = "127.0.0.1"

# Port of the REST proxy
restPort = 9003

# Path to the TLS cert for the gRPC and REST interface
tlsCert = ""

# Path to the TLS private key for the gRPC and REST interface
tlsKey = ""

# Whether the macaroon authentication for the gRPC and REST interface should be disabled
noMacaroons = false

# Path to the admin macaroon for the gRPC and REST interface
adminMacaroonPath = ""

# Path to the read only macaroon for the gRPC and REST interface
readOnlyMacaroonPath = ""
```


CHANGELOG.md

<a name="v1.2.2"></a>
## [v1.2.2] - 2020-12-24
### Fix
- macaroon directory path


<a name="v1.2.1"></a>
## [v1.2.1] - 2020-12-24
### Feat
- add datadir argument
- add Darwin artifacts
- add command to print macaroon in hex
- add readonly macaroon

### Fix
- broken link in gRPC documentation
- set payment fee limit
- use litoshi as denomination for LTC


<a name="v1.2.0"></a>
## [v1.2.0] - 2020-12-14
### Feat
- print parsed config on startup ([#29](https://github.com/BoltzExchange/boltz-lnd/issues/29))
- add network to GetInfo response
- custom paths for REST proxy
- add REST proxy for gRPC server
- add macaroon authentication for gRPC server
- add TLS support for gRPC server

### Fix
- resolve TLS issues of REST proxy


<a name="v1.1.2"></a>
## [v1.1.2] - 2020-12-01
### Fix
- add Swap ID to DepositResponse


<a name="v1.1.1"></a>
## [v1.1.1] - 2020-09-09
### Fix
- nil pointer dereference when LND is killed ([#18](https://github.com/BoltzExchange/boltz-lnd/issues/18))


<a name="v1.1.0"></a>
## [v1.1.0] - 2020-07-29
### Fix
- reconnect to SSE stream
- no LND P2P connection


<a name="v1.0.0"></a>
## v1.0.0 - 2020-07-03
### Feat
- store transaction ids of Reverse Swaps
- store transaction ids of Swaps
- add ListSwaps command ([#15](https://github.com/BoltzExchange/boltz-lnd/issues/15))
- add Channel Creations to GetSwapInfo
- add database schema version
- inbound parameter for deposit call
- prefix for logger
- add withdraw command
- add deposit command
- improve CLI argument parsing
- add lockup address to Reverse Swap response
- add Litecoin support
- handle locked LNDs
- wait for LND to be synced
- add startup sanity checks
- set default Boltz endpoint based on chain
- add Channel Creations
- add Reverse Swaps ([#3](https://github.com/BoltzExchange/boltz-lnd/issues/3))
- fee of refund transactions
- batch refund transactions
- add refund logic
- add Submarine Swaps

### Fix
- show lockup address for Reverse Swaps
- limits of deposit call
- cleanup streams ([#14](https://github.com/BoltzExchange/boltz-lnd/issues/14))
- invoice expiry timeout
- parsing of percentage fee
- bech32 litecoin addresses
- crash in deposit flow
- crash when Swap could not be found
- add Boltz version check
- CLTV encoding

### Refactor
- improve Channel Creation enforcement


[v1.2.2]: https://github.com/BoltzExchange/boltz-lnd/compare/v1.2.1...v1.2.2
[v1.2.1]: https://github.com/BoltzExchange/boltz-lnd/compare/v1.2.0...v1.2.1
[v1.2.0]: https://github.com/BoltzExchange/boltz-lnd/compare/v1.1.2...v1.2.0
[v1.1.2]: https://github.com/BoltzExchange/boltz-lnd/compare/v1.1.1...v1.1.2
[v1.1.1]: https://github.com/BoltzExchange/boltz-lnd/compare/v1.1.0...v1.1.1
[v1.1.0]: https://github.com/BoltzExchange/boltz-lnd/compare/v1.0.0...v1.1.0


config.go
package boltz_lnd

import (
	"fmt"
	"github.com/BoltzExchange/boltz-lnd/boltz"
	"github.com/BoltzExchange/boltz-lnd/build"
	"github.com/BoltzExchange/boltz-lnd/database"
	"github.com/BoltzExchange/boltz-lnd/lnd"
	"github.com/BoltzExchange/boltz-lnd/rpcserver"
	"github.com/BoltzExchange/boltz-lnd/utils"
	"github.com/BurntSushi/toml"
	"github.com/jessevdk/go-flags"
	"os"
	"path"
	"runtime"
)

type helpOptions struct {
	ShowHelp    bool `short:"h" long:"help" description:"Display this help message"`
	ShowVersion bool `short:"v" long:"version" description:"Display version and exit"`
}

type config struct {
	DataDir string `short:"d" long:"datadir" description:"Data directory of boltz-lnd"`

	ConfigFile string `short:"c" long:"configfile" description:"Path to configuration file"`

	LogFile   string `short:"l" long:"logfile" description:"Path to the log file"`
	LogPrefix string `long:"logprefix" description:"Prefix of all log messages"`

	Boltz    *boltz.Boltz         `group:"Boltz Options"`
	LND      *lnd.LND             `group:"LND Options"`
	RPC      *rpcserver.RpcServer `group:"RPC options"`
	Database *database.Database   `group:"Database options"`

	Help *helpOptions `group:"Help Options"`
}

func LoadConfig() *config {
	defaultDataDir, err := utils.GetDefaultDataDir()

	if err != nil {
		fmt.Println("Could not get home directory: " + err.Error())
		os.Exit(1)
	}

	cfg := config{
		DataDir: defaultDataDir,

		ConfigFile: "",

		LogFile:   "",
		LogPrefix: "",

		Boltz: &boltz.Boltz{
			URL: "",
		},

		LND: &lnd.LND{
			Host:        "127.0.0.1",
			Port:        10009,
			Macaroon:    "",
			Certificate: "",
		},

		RPC: &rpcserver.RpcServer{
			Host: "127.0.0.1",
			Port: 9002,

			RestHost:     "127.0.0.1",
			RestPort:     9003,
			RestDisabled: false,

			TlsCertPath: "",
			TlsKeyPath:  "",

			NoMacaroons:          false,
			AdminMacaroonPath:    "",
			ReadonlyMacaroonPath: "",
		},

		Database: &database.Database{
			Path: "",
		},
	}

	parser := flags.NewParser(&cfg, flags.IgnoreUnknown)
	_, err = parser.Parse()

	if cfg.Help.ShowVersion {
		fmt.Println(build.GetVersion())
		fmt.Println("Built with: " + runtime.Version())
		os.Exit(0)
	}

	if cfg.Help.ShowHelp {
		parser.WriteHelp(os.Stdout)
		os.Exit(0)
	}

	if err != nil {
		printCouldNotParse(err)
	}

	cfg.ConfigFile = utils.ExpandDefaultPath(cfg.DataDir, cfg.ConfigFile, "boltz.toml")

	if cfg.ConfigFile != "" {
		_, err := toml.DecodeFile(cfg.ConfigFile, &cfg)

		if err != nil {
			fmt.Printf("Could not read config file: " + err.Error() + "\n")
		}
	}

	_, err = flags.Parse(&cfg)

	if err != nil {
		printCouldNotParse(err)
	}

	fmt.Println("Using data dir: " + cfg.DataDir)

	cfg.LogFile = utils.ExpandDefaultPath(cfg.DataDir, cfg.LogFile, "boltz.log")
	cfg.Database.Path = utils.ExpandDefaultPath(cfg.DataDir, cfg.Database.Path, "boltz.db")

	cfg.RPC.TlsKeyPath = utils.ExpandDefaultPath(cfg.DataDir, cfg.RPC.TlsKeyPath, "tls.key")
	cfg.RPC.TlsCertPath = utils.ExpandDefaultPath(cfg.DataDir, cfg.RPC.TlsCertPath, "tls.cert")

	macaroonDir := path.Join(cfg.DataDir, "macaroons")

	cfg.RPC.AdminMacaroonPath = utils.ExpandDefaultPath(macaroonDir, cfg.RPC.AdminMacaroonPath, "admin.macaroon")
	cfg.RPC.ReadonlyMacaroonPath = utils.ExpandDefaultPath(macaroonDir, cfg.RPC.ReadonlyMacaroonPath, "readonly.macaroon")

	createDirIfNotExists(cfg.DataDir)
	createDirIfNotExists(macaroonDir)

	return &cfg
}

func createDirIfNotExists(dir string) {
	if !utils.FileExists(dir) {
		err := os.Mkdir(dir, 0700)

		if err != nil {
			fmt.Println("Could not create directory: " + err.Error())
			os.Exit(1)
		}
	}
}

func printCouldNotParse(err error) {
	fmt.Println("Could not parse arguments: " + err.Error())
	os.Exit(1)
}
